include/ginkgo/core/factorization/factorization.hpp

```diff
@@ -0,0 +1,165 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_FACTORIZATION_FACTORIZATION_HPP_
+#define GKO_PUBLIC_CORE_FACTORIZATION_FACTORIZATION_HPP_
+
+
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace factorization {
+
+
+enum class storage_type {
```

Comment:
This seems to mix storage details (inplace/composition) with algorithmic details (lu/cholesky). I'm not sure if that is the best way to proceed.

Reply:
I would slightly disagree here: It is necessary to differentiate between different cases (except for composition) for the solver to be able to interpret what to do with the operators. We need to treat the diagonal differently whether it's LDL or Cholesky. We need to know which factors have a unit diagonal to be able to build a solver out of them.
While the name of the algorithm and the factorization match, here I am referring to the latter (LU the factorization, not the algorithm ;) )

Reply:
I think my biggest issue here is the composition. For all other cases, the storage type and algorithm is tightly linked, so I understand these. But the algorithmic detail is completely lost for the composition (except for the symmetry). Of course it's not really relevant for that case, but this makes this enum inconsitent, at least for me.

Side note: perhaps the names `composition_lu`, `composition_cholesky`, `combined_cholesky`, (or `*_llt`), and `combined_ldl` would be better. I think this would help to make it more consistent.

Reply:
In the spirit of avoiding duplication to limit inconsistencies, I thought it would make sense to keep the information "does it involve scaling" in the number of parameters for the composition (2 for no, 3 for yes) and "is it symmetric" in storage_type.

Reply:
There are cases where we need to treat symmetric and non-symmetric factorizations differently, e.g. in a solver like CG expecting a symmetric positive-definite preconditioner, we need to apply a factorization as a two-sided preconditioner, not a normal left-preconditioner. This would even allow us to warn if people attempt to use a non-symmetric factorization preconditioner with a symmetric solver.

Reply:
Yes, of course, but we still have this information in the `cholesky` part.

Reply:
From the name that I've proposed, it would still be clear that the composition is symmetric. So `factor[0] = factor[-1]^T`. But I think I might not understand your issue here.

Reply:
Fair enough, since we already use switch/case for storage_type anyways, it shouldn't matter if we have multiple storage types requiring the same behavior.